pageserver/src/layered_repository/inmemory_layer.rs

```diff
@@ -495,45 +514,31 @@ impl InMemoryLayer {
             tenantid,
             seg,
             start_lsn,
+            end_lsn: None,
             oldest_pending_lsn,
             inner: Mutex::new(InMemoryLayerInner {
                 drop_lsn: None,
                 page_versions: BTreeMap::new(),
                 segsizes,
-                frozen: false,
+                writeable: true,
+                predecessor: Some(src),
             }),
-            predecessor: Some(src),
         })
     }
 
-    ///
-    /// Write the this in-memory layer to disk.
-    ///
-    /// The cutoff point for the layer that's written to disk is 'end_lsn'.
-    ///
-    /// Returns new layers that replace this one. Always returns a new image
-    /// layer containing the page versions at the cutoff LSN, that were written
-    /// to disk, and usually also a DeltaLayer that includes all the WAL records
-    /// between start LSN and the cutoff. (The delta layer is not needed when
-    /// a new relish is created with a single LSN, so that the start and end LSN
-    /// are the same.) If there were page versions newer than 'end_lsn', also
-    /// returns a new in-memory layer containing those page versions. The caller
-    /// replaces this layer with the returned layers in the layer map.
-    ///
-    pub fn freeze(
-        &self,
-        cutoff_lsn: Lsn,
-        // This is needed just to call materialize_page()
-        timeline: &LayeredTimeline,
-    ) -> Result<SuccessorLayers> {
+    /// Splits `self` into two InMemoryLayers: `frozen` and `open`.
+    /// All data up to and including `cutoff_lsn` (or the drop LSN, if dropped)
+    /// is copied to `frozen`, while the remaining data is copied to `open`.
+    /// After completion, self is non-writeable, but not frozen.
```

Comment:
Why do we need to copy data to the new open layer? Can't we just freeze the layer as is?

Reply:
When data has been written since the cutoff I don't want this data to disappear to other readers by making end_lsn smaller (when writeable it's u64::MAX)

Reply:
Hmm, I don't get it.
We receive `last_record_lsn` that is used as `cutoff` under lock and then mark the layer as `non-writeable`. How can any concurrent write since happen after cutoff?